In [1]:
import pandas as pd
import numpy as np
from scipy.stats import f_oneway, chi2_contingency
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, roc_auc_score, log_loss
import os 

In [2]:
pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)

In [3]:
os.getcwd()

'/home/david/Desktop/credit_analysis'

In [4]:
df_current = pd.read_csv("data/processed/df_current_clean", index_col=0)

In [5]:
df_current.head()

,SK_ID_CURR,TARGET,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,REGION_RATING_CLIENT_W_CITY,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAREA_MEDI,NONLIVINGAREA_MEDI,OBS_30_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_YEAR,missing_COMMONAREA_MEDI,missing_NONLIVINGAPARTMENTS_MEDI,missing_FLOORSMIN_MEDI,missing_YEARS_BUILD_MEDI,missing_OWN_CAR_AGE,missing_LANDAREA_MEDI,missing_BASEMENTAREA_MEDI,missing_EXT_SOURCE_1,missing_NONLIVINGAREA_MEDI,missing_ENTRANCES_MEDI,missing_LIVINGAREA_MEDI,missing_FLOORSMAX_MEDI,missing_YEARS_BEGINEXPLUATATION_MEDI,missing_EXT_SOURCE_3,missing_DAYS_EMPLOYED,missing_AMT_REQ_CREDIT_BUREAU_WEEK,missing_AMT_REQ_CREDIT_BUREAU_QRT,missing_AMT_REQ_CREDIT_BUREAU_HOUR,missing_AMT_REQ_CREDIT_BUREAU_MON,missing_AMT_REQ_CREDIT_BUREAU_DAY,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,OCCUPATION_TYPE,WEEKDAY_APPR_PROCESS_START,ORGANIZATION_TYPE,FONDKAPREMONT_MODE,HOUSETYPE_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE,missing_FONDKAPREMONT_MODE,missing_WALLSMATERIAL_MODE,missing_HOUSETYPE_MODE,missing_EMERGENCYSTATE_MODE,missing_OCCUPATION_TYPE
0,100002,1,202500.0,406597.5,24700.5,0.018801,9461.0,637.0,3648.0,2120.0,12.0,1,1,0,1,1,0,1.0,2.0,10.0,0,0,0,0,0.083037,0.262949,0.139376,0.036900,0.972200,0.624300,0.014400,0.069000,0.0833,0.125000,0.037500,0.019300,0.000000,2.0,2.0,1134.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.00000,1.000000,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,Cash loans,M,N,Y,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,Laborers,WEDNESDAY,Business Entity Type 3,reg oper account,block of flats,"Stone, brick",No,0,0,0,0,0
1,100003,0,270000.0,1293502.5,35698.5,0.003541,16765.0,1188.0,1186.0,291.0,11.0,1,1,0,1,1,0,2.0,1.0,11.0,0,0,0,0,0.311267,0.622246,0.512523,0.052900,0.985100,0.798700,0.060800,0.034500,0.2917,0.333300,0.013200,0.055800,0.010000,1.0,0.0,828.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.00000,0.000000,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,Cash loans,F,N,N,Family,State servant,Higher education,Married,House / apartment,Core staff,MONDAY,School,reg oper account,block of flats,Block,No,0,0,0,0,0
2,100004,0,67500.0,135000.0,6750.0,0.010032,19046.0,225.0,4260.0,2531.0,26.0,1,1,1,1,1,0,1.0,2.0,9.0,0,0,0,0,0.649026,0.555912,0.729567,0.059248,0.940569,0.287285,0.023805,0.136941,0.1723,0.184594,0.047017,0.059749,0.029387,0.0,0.0,815.0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.00000,0.000000,1,1,1,1,0,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,Revolving loans,M,Y,Y,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,Laborers,MONDAY,Government,reg oper account,block of flats,Panel,No,1,1,1,1,0
3,100006,0,135000.0,312682.5,29686.5,0.008019,19005.0,3039.0,9833.0,2437.0,13.0,1,1,0,1,0,0,2.0,2.0,17.0,0,0,0,0,0.670032,0.650442,0.512523,0.059248,0.966084,0.595258,0.023805,0.136941,0.1723,0.184594,0.047017,0.059749,0.029387,2.0,0.0,617.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.16961,2.329119,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,1,1,1,1,1,Cash loans,F,N,Y,Unaccompanied,Working,Secondary / seconda

In [6]:
df_previous = pd.read_csv("data/processed/df_previous_clean", index_col=0)

In [7]:
df_previous.head()

,SK_ID_CURR,CODE_REJECT_REASON_CLIENT,CODE_REJECT_REASON_HC,CODE_REJECT_REASON_LIMIT,CODE_REJECT_REASON_SCO,CODE_REJECT_REASON_SCOFR,CODE_REJECT_REASON_SYSTEM,CODE_REJECT_REASON_VERIF,CODE_REJECT_REASON_XAP,CODE_REJECT_REASON_XNA,PRODUCT_COMBINATION_Card Street,PRODUCT_COMBINATION_Card X-Sell,PRODUCT_COMBINATION_Cash,PRODUCT_COMBINATION_Cash Street: high,PRODUCT_COMBINATION_Cash Street: low,PRODUCT_COMBINATION_Cash Street: middle,PRODUCT_COMBINATION_Cash X-Sell: high,PRODUCT_COMBINATION_Cash X-Sell: low,PRODUCT_COMBINATION_Cash X-Sell: middle,PRODUCT_COMBINATION_POS household with interest,PRODUCT_COMBINATION_POS household without interest,PRODUCT_COMBINATION_POS industry with interest,PRODUCT_COMBINATION_POS industry without interest,PRODUCT_COMBINATION_POS mobile with interest,PRODUCT_COMBINATION_POS mobile without interest,PRODUCT_COMBINATION_POS other with interest,PRODUCT_COMBINATION_POS others without interest,NAME_CLIENT_TYPE_New,NAME_CLIENT_TYPE_Refreshed,NAME_CLIENT_TYPE_Repeater,NAME_CLIENT_TYPE_XNA,NAME_CONTRACT_STATUS_Approved,NAME_CONTRACT_STATUS_Canceled,NAME_CONTRACT_STATUS_Refused,NAME_CONTRACT_STATUS_Unused offer,DLDV_median,DLDV_max,CNT_PAYMENT_median,CNT_PAYMENT_max
0,100002,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,1,0,0,0,125.000000,125.00,24.000000,24.0
1,100003,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,0,1,0,1,0,1,0,0,0,0,0,0,2,1,0,3,0,0,0,-647.000000,-386.00,12.000000,12.0
2,100004,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,0,0,0,-694.000000,-694.00,4.000000,4.0
3,100006,0,0,1,0,0,0,0,8,0,1,1,2,0,0,0,1,2,0,1,0,1,0,0,0,0,0,1,0,8,0,5,3,1,0,33767.769531,365243.00,16.054075,48.0
4,100007,0,0,0,0,0,0,0,6,0,0,0,0,1,0,0,0,0,3,1,0,0,0,1,0,0,0,1,0,5,0,6,0,0,0,-535.000000,33767.77,15.000000,48.0


In [8]:
df_previous.shape

(291057, 39)

In [9]:
df_current.shape

(307511, 103)

In [10]:
final_df = df_current.merge(df_previous, on ="SK_ID_CURR", how = "inner")
final_df["TARGET"].value_counts()

TARGET
0    267212
1     23845
Name: count, dtype: int64

i will then build an XGboost model that will predict how risky each customer is,

In [11]:
xgb_params = {
    "objective": "binary:logistic",
    "eval_metric": "logloss",
    "learning_rate": 0.1,      
    "max_depth": 6,               
    "gamma": 0.0,                
    "subsample": 0.8,             
    "colsample_bytree": 0.8,      
    "reg_alpha": 0.1,           
    "reg_lambda": 1.0,            
    "scale_pos_weight": 1.0,      #since the data is not balanced 
    "random_state": 42,
    "n_jobs": -1                  # Use all available CPU cores
}

X = final_df.drop(["TARGET","SK_ID_CURR"], axis=1)
cat_cols = X.select_dtypes(include=['object']).columns
X[cat_cols] = X[cat_cols].astype('category')
y = final_df["TARGET"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=None)

model = XGBClassifier(n_estimators= 700,enable_categorical=True,**xgb_params)
model.fit(X_train, y_train, eval_set=[(X_test, y_test)], verbose=False)

y_probs = model.predict_proba(X_test)[:, 1]

# Convert probabilities to 0 or 1 using a 0.5 threshold
y_preds = (y_probs >= 0.5).astype(int)

# --- METRICS ---
# Standard Accuracy (% of correct predictions)
acc = accuracy_score(y_test, y_preds)
print(f"Accuracy: {acc:.4f}")

Accuracy: 0.9579


In [12]:
final_df["TARGET_probs"] = model.predict_proba(X)[:, 1]
bins = [0.0, 0.3, 0.7, 1.0]
labels = ['low', 'medium', 'high']
# Create the new column
final_df['risk_level'] = pd.cut(final_df['TARGET_probs'], bins=bins, labels=labels, include_lowest=True)


In [13]:
# # trying something 
# final_df = final_df[~(final_df["risk_level"]== "medium")]
# final_df.head()

In [14]:
final_df["risk_level"].unique()

['medium', 'low', 'high']
Categories (3, object): ['low' < 'medium' < 'high']

In [15]:
final_df.head()

,SK_ID_CURR,TARGET,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,CNT_FAM_MEMBERS,REGION_RATING_CLIENT_W_CITY,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,BASEMENTAREA_MEDI,YEARS_BEGINEXPLUATATION_MEDI,YEARS_BUILD_MEDI,COMMONAREA_MEDI,ENTRANCES_MEDI,FLOORSMAX_MEDI,FLOORSMIN_MEDI,LANDAREA_MEDI,LIVINGAREA_MEDI,NONLIVINGAREA_MEDI,OBS_30_CNT_SOCIAL_CIRCLE,DEF_60_CNT_SOCIAL_CIRCLE,DAYS_LAST_PHONE_CHANGE,FLAG_DOCUMENT_2,FLAG_DOCUMENT_3,FLAG_DOCUMENT_4,FLAG_DOCUMENT_5,FLAG_DOCUMENT_6,FLAG_DOCUMENT_7,FLAG_DOCUMENT_8,FLAG_DOCUMENT_9,FLAG_DOCUMENT_10,FLAG_DOCUMENT_11,FLAG_DOCUMENT_12,FLAG_DOCUMENT_13,FLAG_DOCUMENT_14,FLAG_DOCUMENT_15,FLAG_DOCUMENT_16,FLAG_DOCUMENT_17,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_YEAR,missing_COMMONAREA_MEDI,missing_NONLIVINGAPARTMENTS_MEDI,missing_FLOORSMIN_MEDI,missing_YEARS_BUILD_MEDI,missing_OWN_CAR_AGE,missing_LANDAREA_MEDI,missing_BASEMENTAREA_MEDI,missing_EXT_SOURCE_1,missing_NONLIVINGAREA_MEDI,missing_ENTRANCES_MEDI,missing_LIVINGAREA_MEDI,missing_FLOORSMAX_MEDI,missing_YEARS_BEGINEXPLUATATION_MEDI,missing_EXT_SOURCE_3,missing_DAYS_EMPLOYED,missing_AMT_REQ_CREDIT_BUREAU_WEEK,missing_AMT_REQ_CREDIT_BUREAU_QRT,missing_AMT_REQ_CREDIT_BUREAU_HOUR,missing_AMT_REQ_CREDIT_BUREAU_MON,missing_AMT_REQ_CREDIT_BUREAU_DAY,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,OCCUPATION_TYPE,WEEKDAY_APPR_PROCESS_START,ORGANIZATION_TYPE,FONDKAPREMONT_MODE,HOUSETYPE_MODE,WALLSMATERIAL_MODE,EMERGENCYSTATE_MODE,missing_FONDKAPREMONT_MODE,missing_WALLSMATERIAL_MODE,missing_HOUSETYPE_MODE,missing_EMERGENCYSTATE_MODE,missing_OCCUPATION_TYPE,CODE_REJECT_REASON_CLIENT,CODE_REJECT_REASON_HC,CODE_REJECT_REASON_LIMIT,CODE_REJECT_REASON_SCO,CODE_REJECT_REASON_SCOFR,CODE_REJECT_REASON_SYSTEM,CODE_REJECT_REASON_VERIF,CODE_REJECT_REASON_XAP,CODE_REJECT_REASON_XNA,PRODUCT_COMBINATION_Card Street,PRODUCT_COMBINATION_Card X-Sell,PRODUCT_COMBINATION_Cash,PRODUCT_COMBINATION_Cash Street: high,PRODUCT_COMBINATION_Cash Street: low,PRODUCT_COMBINATION_Cash Street: middle,PRODUCT_COMBINATION_Cash X-Sell: high,PRODUCT_COMBINATION_Cash X-Sell: low,PRODUCT_COMBINATION_Cash X-Sell: middle,PRODUCT_COMBINATION_POS household with interest,PRODUCT_COMBINATION_POS household without interest,PRODUCT_COMBINATION_POS industry with interest,PRODUCT_COMBINATION_POS industry without interest,PRODUCT_COMBINATION_POS mobile with interest,PRODUCT_COMBINATION_POS mobile without interest,PRODUCT_COMBINATION_POS other with interest,PRODUCT_COMBINATION_POS others without interest,NAME_CLIENT_TYPE_New,NAME_CLIENT_TYPE_Refreshed,NAME_CLIENT_TYPE_Repeater,NAME_CLIENT_TYPE_XNA,NAME_CONTRACT_STATUS_Approved,NAME_CONTRACT_STATUS_Canceled,NAME_CONTRACT_STATUS_Refused,NAME_CONTRACT_STATUS_Unused offer,DLDV_median,DLDV_max,CNT_PAYMENT_median,CNT_PAYMENT_max,TARGET_probs,risk_level
0,100002,1,202500.0,406597.5,24700.5,0.018801,9461.0,637.0,3648.0,2120.0,12.0,1,1,0,1,1,0,1.0,2.0,10.0,0,0,0,0,0.083037,0.262949,0.139376,0.036900,0.972200,0.624300,0.014400,0.069000,0.0833,0.125000,0.037500,0.019300,0.000000,2.0,2.0,1134.0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.00000,1.000000,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,Cash loans,M,N,Y,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,Laborers,WEDNESDAY,Business Entity Type 3,reg oper account,block of flats,"Stone, brick",No,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0,0,1,0,0,0,125.000000,125.00,24.000000,24.0,0.618592,medium
1,100003,0,270000.0,1293502.5,35698.5,0.003541,16765.0,1188.0,1186.0,291.0,11.0,1,1,0,

### ANALYSIS

## Applicant profile information base on the the risk level:

In [16]:
from IPython.display import display
profil_cols = ["CODE_GENDER","NAME_FAMILY_STATUS","NAME_EDUCATION_TYPE"]

# checking whether there is an association between risk_level and the columns in profil_cols 

def cramers_v(contingency_table):
    chi2 = chi2_contingency(contingency_table)[0]
    n = contingency_table.sum().sum()
    min_dim = min(contingency_table.shape) - 1
    if min_dim == 0:
        return 0
    return np.sqrt(chi2 / (n * min_dim))

dic_list= []
for cols in profil_cols:
    value_table = pd.crosstab(final_df["risk_level"], final_df[cols])
    _, p_value, _, _ = chi2_contingency(value_table)
    c_v = cramers_v(value_table)
    dic_list.append({
    "profil_cols": cols,
    "p_values": p_value,
     "cramers_v": round(c_v, 4)   
    })
result= pd.DataFrame(dic_list).sort_values(by="p_values", ascending= False)

print("p_value for each profile columns : " )
display(result)


info_table1 = pd.get_dummies(final_df[profil_cols]).groupby(final_df["risk_level"], observed= False).sum()
info_table1 = info_table1.reset_index()
info_table1.columns = ["risk_level","females","male","unknown", "civil_marriage", "Married","saparated","single","widows",
                       "academic_degree", "higher_education", "incomplete_higher","lower_secondary", "secondary"]

info_table1 = info_table1.drop("unknown", axis=1)

gender_cols = ["females", "male"]

gender_pct = info_table1[["risk_level"] + gender_cols].copy()

gender_pct[gender_cols] = (
    gender_pct[gender_cols]
    .div(gender_pct[gender_cols].sum(axis=1), axis=0)
    * 100
).round(2)

family_cols = ["civil_marriage","Married","saparated","single","widows"]
family_pct = info_table1[["risk_level"] + family_cols].copy()

family_pct[family_cols] = (
    family_pct[family_cols]
    .div(family_pct[family_cols].sum(axis=1), axis=0)
    * 100
).round(2)

edu_cols = ["academic_degree", "higher_education","incomplete_higher","lower_secondary","secondary"]

edu_pct = info_table1[["risk_level"] + edu_cols].copy()

edu_pct[edu_cols] = (
    edu_pct[edu_cols]
    .div(edu_pct[edu_cols].sum(axis=1), axis=0)
    * 100
).round(2)


p_value for each profile columns : 


,profil_cols,p_values,cramers_v
1,NAME_FAMILY_STATUS,1.082221e-151,0.0353
2,NAME_EDUCATION_TYPE,7.027766e-235,0.0437
0,CODE_GENDER,7.450301e-276,0.0469


In [17]:
gender_pct

,risk_level,females,male
0,low,67.08,32.92
1,medium,56.58,43.42
2,high,53.65,46.35


Female customers constitute the majority across all credit risk categories. However, their representation decreases from 67.08% in the low-risk group to 53.87% in the high-risk group. In contrast, the proportion of male customers increases from 32.92% in the low-risk group to 46.13% in the high-risk group. This pattern indicates that male customers are relatively more represented in the high-risk group, while female customers are relatively more represented in the low-risk group. Overall, gender has a statistically significant but weak association with credit risk (Cramer's V = 0.0469).

In [18]:
edu_pct

,risk_level,academic_degree,higher_education,incomplete_higher,lower_secondary,secondary
0,low,0.06,24.05,3.33,1.23,71.33
1,medium,0.01,16.10,4.25,1.41,78.22
2,high,0.01,12.94,3.03,1.93,82.10


Secondary education is the most common education level across all risk categories, accounting for 71.33% of low-risk customers and increasing to 82.53% among high-risk customers. In contrast, the proportion of customers with higher education declines steadily from 24.05% in the low-risk group to 12.62% in the high-risk group. These findings suggest that customers with higher educational attainment are relatively more represented among low-risk borrowers, whereas customers with secondary education are relatively more represented among high-risk borrowers. The remaining education categories constitute only a small proportion of the customer population and exhibit limited variation across risk levels. Overall, education level has a statistically significant but weak association with credit risk (Cramer's V = 0.0437).

In [19]:
family_pct

,risk_level,civil_marriage,Married,saparated,single,widows
0,low,9.53,64.37,6.43,14.21,5.46
1,medium,11.44,58.75,6.69,19.99,3.13
2,high,13.20,59.42,6.69,17.66,3.03


The difference is not significant enough to show any meaninful information 

Married customers constitute the largest proportion across all credit risk categories, accounting for 64.37% of low-risk customers, 58.75% of medium-risk customers, and 59.42% of high-risk customers. Single customers form the second-largest group, with their representation increasing from 14.21% in the low-risk group to 19.99% in the medium-risk group, before declining slightly to 17.66% in the high-risk group. Customers in civil marriages show a gradual increase in representation as risk increases, rising from 9.53% in the low-risk group to 13.20% in the high-risk group. Separated customers remain relatively stable across risk levels, while widowed customers are less represented in the medium- and high-risk groups than in the low-risk group. Overall, the distribution of family status is relatively similar across risk categories, indicating a weak association between family status and credit risk (Cramer's V = 0.0353).

In [20]:
info_table1[["risk_level","civil_marriage", "Married","saparated","single","widows"]]


,risk_level,civil_marriage,Married,saparated,single,widows
0,low,25835,174477,17439,38514,14806
1,medium,869,4462,508,1518,238
2,high,1635,7363,829,2188,376


## Employment and income base on the risk levels 

In [21]:
Emp_and_inc_cols = ["NAME_INCOME_TYPE", "OCCUPATION_TYPE", "ORGANIZATION_TYPE"]

# checking whether there is an association between risk_level and the columns in Emp_and_inc_cols

dic_list= []
for cols in Emp_and_inc_cols:
    value_table = pd.crosstab(final_df["risk_level"], final_df[cols])
    _, p_value, _, _ = chi2_contingency(value_table)
    c_v = cramers_v(value_table)
    dic_list.append({
    "profil_cols": cols,
    "p_values": p_value,
     "cramers_v": round(c_v, 4)   
    })
result= pd.DataFrame(dic_list).sort_values(by="p_values", ascending= False)

print("p_value for each profile columns : " )
display(result)

info_table2 = pd.get_dummies(final_df[ Emp_and_inc_cols]).groupby(final_df["risk_level"], observed=False).sum()
info_table2 = info_table2.reset_index()
all_cols = info_table2.columns

name_income_type_cols = all_cols[all_cols.str.startswith("NAME_INCOME_TYPE")].str.removeprefix("NAME_INCOME_TYPE_")
name_income_type_cols = name_income_type_cols.str.strip().str.replace(" ","_").to_list()

occupation_type_cols = all_cols[all_cols.str.startswith("OCCUPATION_TYPE")].str.removeprefix("OCCUPATION_TYPE")
occupation_type_cols = occupation_type_cols.str.strip().str.replace(" ", "_").to_list()

organization_type_cols = all_cols[all_cols.str.startswith("ORGANIZATION_TYPE")].str.removeprefix("ORGANIZATION_TYPE")
organization_type_cols = organization_type_cols.str.strip().str.replace(" ", "_").to_list()

info_table2.columns =  ["risk_level"]+name_income_type_cols + occupation_type_cols + organization_type_cols


name_income_type_pct = info_table2[["risk_level"] + name_income_type_cols].copy()
name_income_type_pct[name_income_type_cols]= (name_income_type_pct[name_income_type_cols].div(
    name_income_type_pct[name_income_type_cols].sum(axis=1), axis=0 ) * 100 ).round(2)

occupation_type_pct = info_table2[["risk_level"] + occupation_type_cols].copy()
occupation_type_pct[occupation_type_cols]= (occupation_type_pct[occupation_type_cols].div(
                    occupation_type_pct[occupation_type_cols].sum(axis=1), axis = 0) * 100).round(2)

organization_type_pct = info_table2[["risk_level"] + organization_type_cols].copy()
organization_type_pct[organization_type_cols] = (organization_type_pct[organization_type_cols].div(
     organization_type_pct[organization_type_cols].sum(axis=1), axis = 0) * 100).round(2)


p_value for each profile columns : 


,profil_cols,p_values,cramers_v
1,OCCUPATION_TYPE,2.266801e-256,0.0477
0,NAME_INCOME_TYPE,0.000000e+00,0.0591
2,ORGANIZATION_TYPE,0.000000e+00,0.0656


In [22]:
name_income_type_pct

,risk_level,Commercial_associate,Maternity_leave,Pensioner,State_servant,Student,Unemployed,Working
0,low,22.61,0.00,18.77,7.25,0.0,0.00,51.37
1,medium,21.33,0.00,9.61,5.32,0.0,0.00,63.74
2,high,21.26,0.02,7.84,4.92,0.0,0.06,65.90


The distribution of employment type differs significantly across the credit risk groups (Chi-square test, p < 0.05), although the strength of the association is weak (Cramér's V = 0.059). Working customers are relatively more represented in the high-risk group (65.78%) than in the low-risk group (51.38%), suggesting that working status is associated with a higher credit risk. In contrast, pensioners are more prevalent in the low-risk group (18.75%) than in the high-risk group (7.70%), indicating that pensioner status is associated with a lower credit risk. Commercial associates show relatively stable proportions across the risk groups, suggesting little association with credit risk.

In [23]:
occupation_type_pct

,risk_level,_Accountants,_Cleaning_staff,_Cooking_staff,_Core_staff,_Drivers,_HR_staff,_High_skill_tech_staff,_IT_staff,_Laborers,_Low-skill_Laborers,_Managers,_Medicine_staff,_Private_service_staff,_Realty_agents,_Sales_staff,_Secretaries,_Security_staff,_Waiters/barmen_staff
0,low,4.78,2.22,2.82,13.14,8.69,0.26,5.40,0.24,25.96,0.93,10.15,4.15,1.30,0.36,15.16,0.62,3.19,0.62
1,medium,2.98,2.03,3.34,10.43,9.97,0.22,4.32,0.32,30.19,1.57,7.78,3.70,1.01,0.32,17.10,0.49,3.37,0.87
2,high,2.43,2.67,3.66,9.18,12.02,0.19,3.82,0.18,30.86,2.31,7.00,3.26,1.05,0.27,15.54,0.54,4.19,0.82


The association between occupation type and credit risk is statistically significant (Chi-square test, p < 0.05), although the strength of the association is very weak (Cramér's V = 0.04). Laborers represent the largest occupational group across all risk categories, with their proportion increasing from 25.94% in the low-risk group to 31.05% in the high-risk group. Similarly, the proportion of drivers increases from 8.68% to 11.95% as risk increases. In contrast, accountants, managers, core staff, and high-skill technical staff are relatively more represented in the low-risk group. While these patterns suggest differences in the occupational composition of the risk groups, the small Cramér's V indicates that occupation type alone has limited ability to distinguish between low- and high-risk customers.

In [24]:
organization_type_pct 

,risk_level,_Advertising,_Agriculture,_Bank,_Business_Entity_Type_1,_Business_Entity_Type_2,_Business_Entity_Type_3,_Cleaning,_Construction,_Culture,_Electricity,_Emergency,_Government,_Hotel,_Housing,_Industry:_type_1,_Industry:_type_10,_Industry:_type_11,_Industry:_type_12,_Industry:_type_13,_Industry:_type_2,_Industry:_type_3,_Industry:_type_4,_Industry:_type_5,_Industry:_type_6,_Industry:_type_7,_Industry:_type_8,_Industry:_type_9,_Insurance,_Kindergarten,_Legal_Services,_Medicine,_Military,_Mobile,_Other,_Police,_Postal,_Realtor,_Religion,_Restaurant,_School,_Security,_Security_Ministries,_Self-employed,_Services,_Telecom,_Trade:_type_1,_Trade:_type_2,_Trade:_type_3,_Trade:_type_4,_Trade:_type_5,_Trade:_type_6,_Trade:_type_7,_Transport:_type_1,_Transport:_type_2,_Transport:_type_3,_Transport:_type_4,_University,_XNA
0,low,0.14,0.79,0.78,1.89,3.38,21.75,0.08,2.08,0.13,0.30,0.18,3.42,0.32,0.95,0.32,0.04,0.86,0.12,0.02,0.15,1.04,0.28,0.19,0.04,0.42,0.01,1.10,0.18,2.30,0.09,3.72,0.89,0.10,5.46,0.78,0.71,0.12,0.03,0.58,2.99,1.04,0.66,12.44,0.52,0.18,0.11,0.52,0.99,0.02,0.02,0.21,2.49,0.07,0.72,0.36,1.72,0.44,18.77
1,medium,0.08,0.70,0.68,2.11,4.42,25.06,0.04,3.37,0.07,0.29,0.17,2.80,0.21,0.92,0.57,0.04,1.17,0.03,0.03,0.13,1.42,0.28,0.18,0.05,0.43,0.01,1.15,0.11,2.34,0.07,2.92,0.58,0.11,4.91,0.53,0.99,0.14,0.01,0.75,1.79,1.54,0.46,16.00,0.46,0.18,0.13,0.59,1.63,0.01,0.00,0.14,3.54,0.03,0.82,0.67,2.25,0.26,9.61
2,high,0.17,1.44,0.45,1.88,3.43,24.99,0.19,3.74,0.11,0.29,0.17,3.11,0.31,1.09,0.52,0.02,1.02,0.06,0.06,0.16,1.67,0.41,0.17,0.02,0.46,0.02,0.95,0.12,1.72,0.10,2.95,0.51,0.10,5.40,0.45,0.82,0.19,0.03,1.05,2.34,1.40,0.35,17.36,0.46,0.24,0.14,0.51,1.28,0.00,0.01,0.13,3.40,0.05,0.64,1.02,2.17,0.27,7.90


Organization type is significantly associated with credit risk (Chi-square test, p < 0.05); however, the strength of the association is weak (Cramér's V = 0.062). Business Entity Type 3 and self-employed organizations account for a larger proportion of customers in the high-risk group, whereas the proportion of customers with an unknown organization type (XNA) and those employed in the medical sector decreases as credit risk increases. Most of the remaining organization categories exhibit only minor differences across the risk groups, indicating that organization type alone has limited discriminatory power for identifying high-risk customers.

## Asset and wealth base on the risk levels

In [25]:
asset_and_wealth_cols = ["FLAG_OWN_CAR", "FLAG_OWN_REALTY", "NAME_HOUSING_TYPE"]

# checking whether there is an association between risk_level and the columns in asset_and_wealth_cols

dic_list= []
for cols in asset_and_wealth_cols:
    value_table = pd.crosstab(final_df["risk_level"], final_df[cols])
    _, p_value, _, _ = chi2_contingency(value_table)
    c_v = cramers_v(value_table)
    dic_list.append({
    "profil_cols": cols,
    "p_values": p_value,
     "cramers_v": round(c_v, 4)   
    })
result= pd.DataFrame(dic_list).sort_values(by="p_values", ascending= False)

print("p_value for each profile columns : " )
display(result)

info_table2 = pd.get_dummies(final_df[asset_and_wealth_cols]).groupby(final_df["risk_level"], observed=False).sum()
info_table2 = info_table2.reset_index()
all_cols = info_table2.columns

flag_own_car_cols = all_cols[all_cols.str.startswith("FLAG_OWN_CAR")].str.removeprefix("FLAG_OWN_CAR_")
flag_own_car_cols = flag_own_car_cols.str.strip().str.replace("","1").to_list()

flag_own_realty_cols = all_cols[all_cols.str.startswith("FLAG_OWN_REALTY")].str.removeprefix("FLAG_OWN_REALTY_")
flag_own_realty_cols = flag_own_realty_cols.str.strip().str.replace(" ","_").to_list()


name_housing_cols = all_cols[all_cols.str.startswith("NAME_HOUSING_TYPE")].str.removeprefix("NAME_HOUSING_TYPE_")
name_housing_cols = name_housing_cols.str.strip().str.replace(" ","_").to_list()

info_table2.columns =  ["risk_level"] + flag_own_car_cols + flag_own_realty_cols + name_housing_cols

flag_own_car_pct = info_table2[["risk_level"] + flag_own_car_cols].copy()
flag_own_car_pct[flag_own_car_cols]= (flag_own_car_pct[flag_own_car_cols].div(
    flag_own_car_pct[flag_own_car_cols].sum(axis=1), axis=0 ) * 100 ).round(2)

flag_own_realty_pct = info_table2[["risk_level"] + flag_own_realty_cols].copy()
flag_own_realty_pct[flag_own_realty_cols] = (flag_own_realty_pct[flag_own_realty_cols].div(
    flag_own_realty_pct[flag_own_realty_cols].sum(axis=1), axis = 0) * 100).round(2) 

name_housing_pct = info_table2[["risk_level"] + name_housing_cols].copy()
name_housing_pct[name_housing_cols] = (name_housing_pct[name_housing_cols].div(
    name_housing_pct[name_housing_cols].sum(axis=1), axis = 0) * 100).round(2) 
                                     


p_value for each profile columns : 


,profil_cols,p_values,cramers_v
1,FLAG_OWN_REALTY,1.833944e-12,0.0136
0,FLAG_OWN_CAR,1.792838e-30,0.0217
2,NAME_HOUSING_TYPE,2.805465e-121,0.0320


In [26]:
flag_own_car_pct

,risk_level,1N1,1Y1
0,low,65.89,34.11
1,medium,68.77,31.23
2,high,70.51,29.49


Customers who do not own a car constitute the majority across all risk levels. Their proportion increases slightly from 65.88% in the low-risk group to 70.65% in the high-risk group, while the proportion of customers who own a car decreases from 34.12% to 29.35%.

The chi-square test indicates that the association between car ownership and risk level is statistically significant (p = 1.828 × 10⁻³²), suggesting that the observed differences are unlikely to have occurred by random chance. However, the strength of this association is very weak, as indicated by a Cramér's V value of 0.022. Therefore, although customers without a car are slightly more represented in the high-risk group and car owners are relatively more common in the low-risk group, car ownership has limited practical importance as a predictor of credit risk

In [27]:
flag_own_realty_pct

,risk_level,N,Y
0,low,29.19,70.81
1,medium,33.03,66.97
2,high,29.78,70.22


Although the chi-square test indicated a statistically significant association between real estate ownership and risk level (p < 0.001), the association was negligible (Cramér's V = 0.0152). The proportions of customers owning real estate were nearly identical across risk levels, suggesting that real estate ownership has little practical relevance for distinguishing customer risk."

In [28]:
name_housing_pct

,risk_level,Co-op_apartment,House_/_apartment,Municipal_apartment,Office_apartment,Rented_apartment,With_parents
0,low,0.36,88.94,3.68,0.86,1.52,4.63
1,medium,0.49,82.88,3.78,0.72,2.94,9.19
2,high,0.32,86.03,4.10,0.67,2.40,6.47


Housing type was found to be statistically associated with risk level (p = 1.628 × 10⁻¹³⁹). However, the strength of the association was very weak (Cramér's V = 0.0342), indicating limited practical importance.

Across all risk groups, the majority of customers lived in a house or apartment, accounting for over 82% of customers in each group. The most notable difference was observed among customers living with their parents, whose proportion was highest in the medium-risk group (9.95%) compared with the low-risk (4.61%) and high-risk (6.44%) groups. Other housing categories showed only minor differences across the risk levels.

Overall, although housing type is statistically associated with credit risk, the weak effect size suggests that it is not a strong discriminator of customer risk.

## Loan Characteristics base on the risk level 

In [36]:
loan_xteristics_cols = ["NAME_CONTRACT_TYPE", "CNT_PAYMENT_median","DLDV_median"]

# Helper for Eta-Squared (Effect size for ANOVA / continuous variables)
def eta_squared(groups):
    all_data = np.concatenate(groups)
    grand_mean = np.mean(all_data)
    sst = np.sum((all_data - grand_mean) ** 2)
    ssb = np.sum(
        [len(g) * (np.mean(g) - grand_mean) ** 2 for g in groups if len(g) > 0]
    )
    return ssb / sst if sst != 0 else 0


# Statistical Tests & Association Table ---
dic_list = []

for cols in loan_xteristics_cols:
    if final_df[cols].dtype == "object" or final_df[cols].nunique() <= 10:
        # Categorical feature -> Chi-Square Test & Cramer's V
        value_table = pd.crosstab(final_df["risk_level"], final_df[cols])
        _, p_value, _, _ = chi2_contingency(value_table)
        effect_size = round(cramers_v(value_table), 4)
        stat_metric = "Cramer's V"
    else:
        # Numerical/Continuous feature -> One-Way ANOVA & Eta-Squared
        groups = [
            group[cols].dropna().values
            for _, group in final_df.groupby("risk_level", observed=False)
        ]
        _, p_value = f_oneway(*groups)
        effect_size = round(eta_squared(groups), 4)
        stat_metric = "Eta-Squared"

    dic_list.append(
        {
            "profil_cols": cols,
            "p_values": p_value,
            "effect_size": effect_size,
            "effect_type": stat_metric,
        }
    )

result = pd.DataFrame(dic_list).sort_values(by="p_values", ascending=False)
print("Statistical summary for loan characteristics columns:")
display(result)



contract_type_pct = (
    pd.crosstab(
        final_df["risk_level"], final_df["NAME_CONTRACT_TYPE"], normalize="index"
    )
    * 100
).round(2)
contract_type_pct = contract_type_pct.reset_index()

#  Numerical: CNT_PAYMENT & DLDV_median (Mean/Median by risk_level)
loan_num_stats = (
    final_df.groupby("risk_level", observed= False)[["CNT_PAYMENT_median", "DLDV_median"]]
    .agg(["mean"])
    .round(2)
)
loan_num_stats.columns = [
    f"{col}_{stat}" for col, stat in loan_num_stats.columns
]
loan_num_stats = loan_num_stats.reset_index()
    

Statistical summary for loan characteristics columns:


,profil_cols,p_values,effect_size,effect_type
1,CNT_PAYMENT_median,4.153058e-49,0.0008,Eta-Squared
0,NAME_CONTRACT_TYPE,5.354925e-66,0.0321,Cramer's V
2,DLDV_median,1.535864e-116,0.0018,Eta-Squared


In [37]:
print("\n--- NAME_CONTRACT_TYPE Percentage Distribution ---")
display(contract_type_pct)



--- NAME_CONTRACT_TYPE Percentage Distribution ---


NAME_CONTRACT_TYPE,risk_level,Cash loans,Revolving loans
0,low,90.68,9.32
1,medium,94.42,5.58
2,high,94.28,5.72


Cash loans are predominant across all risk levels, accounting for over 90% of contracts. Their proportion increases slightly from 90.68% among low-risk customers to 94.28% among high-risk customers, while revolving loans decline correspondingly. However, the small difference and very weak Cramér's V (0.0321) indicate that contract type has limited practical importance in distinguishing customer risk.

In [38]:
print("\n--- Numerical Loan Characteristics Stats by Risk Level ---")
display(loan_num_stats)


--- Numerical Loan Characteristics Stats by Risk Level ---


,risk_level,CNT_PAYMENT_median_mean,DLDV_median_mean
0,low,13.49,14729.28
1,medium,14.34,21234.69
2,high,14.16,20718.32


The average number of payments is relatively similar across the risk levels, with no clear trend from low to high risk. In contrast, DLDV is substantially higher among medium- and high-risk customers compared with the low-risk group, suggesting that loan due-date characteristics differ more noticeably across risk levels than the number of payments.